src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js

```diff
@@ -0,0 +1,533 @@
+            this.bins        = data.bins         || this.bins;
+        } else {
+            this.axisSpec    = data.axis;
+            this.focus       = data.focus;
+            this.dataValues  = data.data.values;
+            this.bins        = data.bins;
+        }
+    }
+
+    /**
+     * Return the focus area of the histogram. If none is set, returns undefined.
+     */
+    getFocus() {
+        return this.focus
+    }
+
+    /**
+     * Return specification for the axis. This includes scales (logarithmic/linear) and labels.
+     */
+    getAxisSpec() {
+        return this.axisSpec || {x:{scale:LINEAR_SCALE},y:{scale:LINEAR_SCALE}}
+    }
+
+    /**
+     * Return vales to plot.
+     */
+    getData() {
+        return this.dataValues || {};
+    }
+
+    /**
+     * Return the number of bins to use for the histogram..
+     */
+    getBinCount() {
+        return this.bins || DEFAULT_NUMBER_OF_BINS
+    }
+
+    /**
+     * Return the layout measurements for the plot. This includes the outer dimensions of the
+     * drawing area as well as the inner dimensions of the plotting area and the margins.
+     */
+    getCanvasDimensions() {
+        const width  = this.dom.getAttributeNS(null,"width");
+        let height   = this.dom.getAttributeNS(null,"height");
+        height       = height - BUTTON_HEIGHT;
+        const margin = this.getMargins(this.getAxisSpec());
+        return {
+            inner : {
+                width  : width - margin.left - margin.right,
+                height : height - margin.top - margin.bottom
+            },
+            outer : {
+                width  : width,
+                height : height,
+            },
+            margin : margin,
+        };
+    }
+
+    /**
+     * Update the histogram with the current data and settings.
+     */
+    updateHistogram() {
+        const canvas  = this.getCanvasDimensions();
+        const divElem = this.createDivElem(canvas.outer.width,canvas.outer.height);
+        this.dom.appendChild(divElem);
+
+        const svg = d3.select(this.dom)
+            .append("svg")
+            .attr("width",canvas.outer.width)
+            .attr("height",canvas.outer.height)
+            .append("g")
+            .attr("transform","translate(" + canvas.margin.left + "," + canvas.margin.top + ")");
+
+        const extremesAndDeltas = this.getExtremesAndDeltas(this.getData());
+        this.createLabels(this.getAxisSpec(),svg,canvas);
+        const [bars,scale,axis] = this.createHistogram(extremesAndDeltas,canvas,svg,this.getData(),this.getFocus());
+        const zoom = this.initPanAndZoom(canvas,svg,bars,scale,axis);
+
+        // TODO [MM]: In task specification buttons were on top of the visualization, but because
+        //            the visualization selector obfuscated them, they're now on the bottom.
+        //            This should be fixed in (#898).
+        this.createButtonFitAll(scale,axis,svg,bars,extremesAndDeltas,canvas,zoom);
```

Comment:
Just a fast question - I really, really rarely see in OOP functions with so many arguments. Are you sure we dont want to store some of them as class fields?

Reply:
You are right. I did not completely refactor this before, but it is changed now.